tools/image-info

```diff
@@ -407,20 +424,499 @@ def read_sysconfig(tree):
+    return result
+
+
+# Read all uncommented key/values set in /etc/systemd/logind.conf
+def read_logind_conf(tree):
+    result = {}
+
+    with contextlib.suppress(FileNotFoundError):
+        with open(f"{tree}/etc/systemd/logind.conf") as f:
+            parser = configparser.RawConfigParser()
+            # prevent conversion of the opion name to lowercase
+            parser.optionxform = lambda option: option
+            parser.read_file(f)
+            with contextlib.suppress(configparser.NoSectionError):
+                result.update(parser["Login"])
+    return result
+
+
+# Read all uncommented key/values set in /etc/locale.conf
+def read_locale(tree):
+    with contextlib.suppress(FileNotFoundError):
+        with open(f"{tree}/etc/locale.conf") as f:
+            return parse_environment_vars(f.read())
+
+
+# Read information related to SELinux
+def read_selinux_info(tree):
+    result = {}
+
+    policy = read_selinux_conf(tree)
+    if policy:
+        result["policy"] = policy
+
+    ctx_mismatch = read_selinux_ctx_mismatch(tree)
+    if ctx_mismatch:
+        result["context-mismatch"] = ctx_mismatch
+
+    return result
+
+
+# Read all uncommented key/values set in /etc/selinux/config
+def read_selinux_conf(tree):
+    with contextlib.suppress(FileNotFoundError):
+        with open(f"{tree}/etc/selinux/config") as f:
+            return parse_environment_vars(f.read())
+
+
+# Read any mismatch in selinux context of files on the image
+def read_selinux_ctx_mismatch(tree):
+    result = []
+
+    # The binary policy that should be used is on the image and has name "policy.X"
+    # where the "X" is a number. There may be more than one policy files.
+    # In the usual case, the policy with the highest number suffix should be used.
+    policy_files = glob.glob(f"{tree}/etc/selinux/targeted/policy/policy.*")
+    policy_files = sorted(policy_files, reverse=True)
+
+    if policy_files:
+        CMD = [
+            "setfiles",
+            "-r", f"{tree}",
+            "-nvF",
+            "-c", policy_files[0],  # take the policy with the highest number
+            f"{tree}/etc/selinux/targeted/contexts/files/file_contexts",
+            f"{tree}"
+        ]
+
+        output = subprocess.check_output(CMD).decode()
+
+        # output are lines such as:
+        # Would relabel /tmp/tmpwrozmb47/dev from system_u:object_r:root_t:s0 to system_u:object_r:device_t:s0\n
+        setfiles_pattern = r"Would\s+relabel\s+(?P<filename>.+)\s+from\s+(?P<actual>.+)\s+to\s+(?P<expected>.+)"
+        setfiles_re = re.compile(setfiles_pattern)
+
+        for line in output.splitlines():
+            line = line.strip()
+            if not line:
+                continue
+            match = setfiles_re.match(line)
+            # do not silently ignore changes of 'setfiles' output
+            if not match:
+                raise RuntimeError("could not match line '{}' with pattern '{}'".format(line, setfiles_pattern))
```

Comment:
f-string? :)

Reply:
Oh yes, I'm just still not used to using it :smile: